### DISCUSSION TITLE
Main module initialization order

### DISCUSSION THREAD

**Discussion by allsey87** [emscripten-core/emscripten#19695] [Q&A] [answered]
The main module's constructor seems to be called by the runtime before its thread local storage has been initialized. Should this be the case? Perhaps it is just my ignorance, but this seems a bit off to me.

**Comment by sbc100** [comment]
That certainly should not be the case no.    TLS initialization should happen before any code (including static constructors) run.   

Perhaps I'm not understanding correctly.. what do you mean by "main module's constructor"

**Reply by allsey87** [reply]
Hi Sam, by main module's constructor, I mean `Module["asm"]["__wasm_call_ctors"]` seems to be getting called before `Module["asm"]["_emscripten_tls_init"]` (based on wrapping the two in a function that also prints to `Console.log`).

**Reply by allsey87** [reply]
@sbc100 do you also observe this on your end?

**Reply by sbc100** [reply]
TLS init is performed by the `PThread.threadInitTLS()` function.

For the main thread this is called by `__emscripten_init_main_thread_js`, which itself is called from a static constructor using a system level priority to put it first above all user static constructor: https://github.com/emscripten-core/emscripten/blob/68ae550a0a6d64454e306c27fa72abbf173de773/system/lib/pthread/library_pthread.c#L576-L579

The list of system-level static constructors is here: 
https://github.com/emscripten-core/emscripten/blob/68ae550a0a6d64454e306c27fa72abbf173de773/system/lib/README.md?plain=1#L19-L24

**Reply by sbc100** [reply, marked as answer]
In order works `__wasm_call_ctors` is the thing that actually calls `_emscripten_tls_init` .. but this is by design.
